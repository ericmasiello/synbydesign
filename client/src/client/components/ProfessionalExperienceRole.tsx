import * as React from 'react';
import styled from 'styled-components';
import TypeBase from './TypeBase';
import TypeSmall from './TypeSmall';
import Muted from './Muted';
import { TYPE_SIZE, COLORS } from '../styles/vars';
import { pxToRem } from '../styles/utils';

const TIMELINE_IN_MARGIN = '115rem';

export interface Props extends ProfessionalRole {
  className?: string;
}

const computeLineHeight = (size: TypeSize) => {
  return pxToRem(size[0] * size[1]);
};

const getTimeline = (yearFrom: string, yearTo: string = 'Present') => (
  <TypeSmall tag={Muted} className="role__timeline">
    <span className="role__parentheses">(</span>
    {yearFrom} &mdash; {yearTo}
    <span className="role__parentheses">)</span>
  </TypeSmall>
);

export const ProfessionalExperienceRole: React.SFC<Props> = props => {
  const { className, title, yearFrom, yearTo, ...rest } = props;
  return (
    <div className={className} {...rest}>
      <TypeBase tag="h2" className="role__title">
        {title} {getTimeline(yearFrom, yearTo)}
      </TypeBase>
    </div>
  );
};

ProfessionalExperienceRole.displayName = 'ProfessionalExperience.Role';

export default styled(ProfessionalExperienceRole)`
  position: relative;

  .role__title {
    color: ${COLORS.highlight};
    margin-bottom: 0.25rem;
  }

  &:nth-last-of-type(1) {
    .role__title {
      margin-bottom: 0.5rem;
    }
  }

  .role__parentheses {
    @media (min-width: ${TIMELINE_IN_MARGIN}) {
      display: none;
    }
  }

  .role__timeline {
    @media (min-width: ${TIMELINE_IN_MARGIN}) {
      position: absolute;
      top: 0;
      left: -1rem;
      transform: translateX(-100%);
      line-height: ${computeLineHeight(TYPE_SIZE.base)};
    }
  }
`;
