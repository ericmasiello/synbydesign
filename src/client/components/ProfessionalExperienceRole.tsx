import * as React from 'react';
import styled from 'styled-components';
import TypeBase from './TypeBase';
import TypeSmall from './TypeSmall';
import { TYPE_SIZE, COLORS } from '../styles/vars';
import { pxToRem } from '../styles/utils';

export interface Props extends ProfessionalRole {
  className?: string;
}

const computeLineHeight = (size: TypeSize) => {
  return pxToRem(size[0] * size[1]);
};

const getTimeline = (yearFrom: string, yearTo: string = 'Present') => (
  <TypeSmall tag="p" className="role__timeline">
    {yearFrom} &mdash; {yearTo}
  </TypeSmall>
);

export const ProfessionalExperienceRole: React.SFC<Props> = props => {
  const { className, title, yearFrom, yearTo, ...rest } = props;
  return (
    <div className={className} {...rest}>
      <TypeBase tag="h2" className="role__title">
        {title}
      </TypeBase>
      {getTimeline(yearFrom, yearTo)}
    </div>
  );
};

ProfessionalExperienceRole.displayName = 'ProfessionalExperience.Role';

export default styled(ProfessionalExperienceRole)`
  position: relative;

  .role__title {
    color: ${COLORS.highlight};
    margin-bottom: 0.5rem;
  }

  .role__timeline {
    position: absolute;
    top: 0;
    left: -1rem;
    transform: translateX(-100%);
    line-height: ${computeLineHeight(TYPE_SIZE.base)};
  }
`;
