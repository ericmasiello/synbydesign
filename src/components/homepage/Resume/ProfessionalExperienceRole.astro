---
import classNames from 'classnames';
import type { Role } from './Resume';
import Text from '../../Text.astro';
import Timeline from './Timeline.astro';

export interface Props extends Omit<astroHTML.JSX.BaseHTMLAttributes, 'slot' | 'title'>, Role {
  as?: keyof HTMLElementTagNameMap;
}

const { as: Element = 'div', class: className, title, yearFrom, yearTo, ...rest } = Astro.props;
const classes = classNames('role', className);
---

<Element class={classes} {...rest}>
  <Text as="h5" class="title">
    {title}{' '}
    <Timeline class="timeline" parenthesesClass="parentheses">
      {yearFrom} &mdash; {yearTo ?? 'Present'}
    </Timeline>
  </Text>
</Element>

<style>
  .role {
    position: relative;
  }

  .title {
    color: var(--color-brand);
    margin-bottom: 0.25rem;
  }

  .role:nth-last-of-type(1) .title {
    margin-bottom: 0.5rem;
  }

  @media (min-width: 115rem) {
    .timeline {
      position: absolute;
      top: 0;
      left: -1rem;
      transform: translateX(-100%);
      line-height: 1.6;
    }
  }

  @media (min-width: 115rem) {
    .parentheses {
      display: none;
    }
  }
</style>
