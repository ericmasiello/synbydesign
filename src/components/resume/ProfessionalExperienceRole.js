import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Text } from '../Type';
import Timeline from './Timeline';
import * as styles from './ProfessionalExperienceRole.module.css';

function ProfessionalExperienceRole(props) {
  const { as: Component, className, title, yearFrom, yearTo, ...rest } = props;
  const classes = classNames(styles.role, className);
  return (
    <Component className={classes} {...rest}>
      <Text as="h5" className={styles.title}>
        {title}{' '}
        <Timeline className={styles.timeline} parenthesesClassName={styles.parentheses}>
          {yearFrom} &mdash; {yearTo ?? 'Present'}
        </Timeline>
      </Text>
    </Component>
  );
}

ProfessionalExperienceRole.defaultProps = {
  as: 'div',
};

ProfessionalExperienceRole.propTypes = {
  as: PropTypes.elementType,
};

export default ProfessionalExperienceRole;
