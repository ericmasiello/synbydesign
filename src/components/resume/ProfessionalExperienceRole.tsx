import React from 'react';
import classNames from 'classnames';
import { Text } from '../Type';
import { Timeline } from './Timeline';
import type { ElementType } from 'react';
import type { FlexibleComponent } from '../../@types/FlexibleComponent';
import * as styles from './ProfessionalExperienceRole.module.css';

type Props = {
  className?: string;
  title: string;
  yearFrom: string;
  yearTo?: string;
};

export const ProfessionalExperienceRole = <T extends ElementType = 'div'>(props: FlexibleComponent<T, Props>) => {
  const { component: Component = 'div', className, title, yearFrom, yearTo, ...rest } = props;
  const classes = classNames(styles.role, className);
  return (
    <Component className={classes} {...rest}>
      <Text component="h5" className={styles.title}>
        {title}{' '}
        <Timeline className={styles.timeline} parenthesesClassName={styles.parentheses}>
          {yearFrom} &mdash; {yearTo ?? 'Present'}
        </Timeline>
      </Text>
    </Component>
  );
};
