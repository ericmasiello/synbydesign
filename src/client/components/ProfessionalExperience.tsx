import * as React from 'react';
import styled from 'styled-components';
import Tag from './Tag';
import ProfessionalExperienceRole from './ProfessionalExperienceRole';

interface Props extends ProfessionalExperience {
  className?: string;
  tag?: Tag;
}

interface DefaultProps {
  tag: Tag;
  accomplishments: string[];
}

export const ProfessionalExperience: React.SFC<Props> = props => {
  const {
    organization,
    roles,
    accomplishments,
    className,
    ...rest
  } = props as Props & DefaultProps;
  return (
    <Tag className={className} {...rest}>
      <hgroup>
        <h1>{organization}</h1>
        <ProfessionalExperienceRole roles={roles} />
      </hgroup>
      {accomplishments.length > 0 && (
        <section>
          <ul>
            {accomplishments.map(accomplishment => (
              <li key={accomplishment}>{accomplishment}</li>
            ))}
          </ul>
        </section>
      )}
    </Tag>
  );
};

ProfessionalExperience.defaultProps = {
  tag: 'article',
  accomplishments: [],
} as DefaultProps;

ProfessionalExperience.displayName = 'ProfessionalExperience';

export default styled(ProfessionalExperience)``;
