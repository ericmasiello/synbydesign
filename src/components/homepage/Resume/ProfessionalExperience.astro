---
import classNames from 'classnames';
import H from '../../H.astro';
import PlainList from '../../PlainList.astro';
import ProfessionalExperienceRole from './ProfessionalExperienceRole.astro';
import type { ProfessionalExperience } from './Resume';
import Spacer from './Spacer.astro';
import type { Props as SpacerProps } from './Spacer.astro';

export interface Props extends Omit<SpacerProps, 'slot'>, ProfessionalExperience {}

const { as = 'article', class: className, organization, roles, accomplishments, ...rest } = Astro.props;
const classes = classNames('experience', className);
---

<Spacer class={classes} {...rest} as={as}>
  <hgroup>
    <H as="h4" class="org" bold>
      {organization}
    </H>
    <PlainList class="roles" spacing="small">{roles.map((role) => <ProfessionalExperienceRole {...role} />)}</PlainList>
  </hgroup>
  {
    accomplishments.length > 0 && (
      <div>
        <PlainList>
          {accomplishments.map((accomplishment) => (
            <li>{accomplishment}</li>
          ))}
        </PlainList>
      </div>
    )
  }
</Spacer>
<style>
  .org {
    margin-block-end: 0.5rem;
  }

  .roles {
    margin-block-end: 0.5rem;
  }
</style>
