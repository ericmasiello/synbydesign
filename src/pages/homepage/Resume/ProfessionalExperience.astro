---
import classNames from 'classnames';
import H from '../../../components/H.astro';
import PlainList from '../../../components/PlainList.astro';
import ProfessionalExperienceRole from './ProfessionalExperienceRole.astro';
import type { ProfessionalExperience } from './Resume';

interface Props extends Omit<astroHTML.JSX.BaseHTMLAttributes, 'slot'>, ProfessionalExperience {
  as?: keyof HTMLElementTagNameMap;
}

const { as: Element = 'hgroup', class: className, organization, roles, accomplishments, ...rest } = Astro.props;
const classes = classNames('experience', className);
---

<Element class={classes} {...rest}>
  <hgroup>
    <H as="h4" class="org">
      {organization}
    </H>
    {roles.map((role) => <ProfessionalExperienceRole {...role} />)}
  </hgroup>
  {
    accomplishments.length > 0 && (
      <div>
        <PlainList>
          {accomplishments.map((accomplishment) => (
            <li>{accomplishment}</li>
          ))}
        </PlainList>
      </div>
    )
  }
</Element>
