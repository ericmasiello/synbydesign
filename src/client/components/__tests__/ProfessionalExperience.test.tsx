import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { ProfessionalExperience } from '../ProfessionalExperience';
import ProfessionalExperienceRole from '../ProfessionalExperienceRole';

const base: ProfessionalExperience = {
  organization: 'Company Name',
  roles: [],
};

it('should render', () => {
  const component = renderer.create(<ProfessionalExperience {...base} />);
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});

it('should display additional classes', () => {
  const wrapper = shallow(
    <ProfessionalExperience {...base} className="my-class" />,
  );

  expect(wrapper.hasClass('my-class')).toBe(true);
});

it('should render the organization', () => {
  const wrapper = shallow(
    <ProfessionalExperience {...base} className="my-class" />,
  );

  expect(wrapper.find('h1').text()).toBe(base.organization);
});

it('should render roles', () => {
  const roles: ProfessionalRole[] = [
    {
      title: 'Title 1',
      yearFrom: '2001',
      yearTo: '2020',
    },
  ];
  const wrapper = shallow(<ProfessionalExperience {...base} roles={roles} />);
  const professionalExperience = wrapper.find(ProfessionalExperienceRole);

  expect(professionalExperience.props().roles).toEqual(roles);
});

it('should render accomplishments', () => {
  const accomplishments = ['foo', 'bar', 'baz'];
  const wrapper = shallow(
    <ProfessionalExperience {...base} accomplishments={accomplishments} />,
  );

  accomplishments.forEach(accomplishment => {
    expect(wrapper.containsMatchingElement(<li>{accomplishment}</li>)).toBe(
      true,
    );
  });
});
