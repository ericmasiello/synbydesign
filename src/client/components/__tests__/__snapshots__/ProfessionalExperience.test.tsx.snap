// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<article
  className={undefined}
>
  <hgroup>
    <h1>
      Company Name
    </h1>
  </hgroup>
</article>
`;
